test/Elastic.Apm.Tests.Utilities/MockPayloadSender.cs

```diff
@@ -72,16 +69,22 @@ public bool WaitForTransactions(TimeSpan? timeout = null, int? count = null)
 				var signalled = true;
 				if (timeout is null)
 				{
-					while (_transactions.Count < count && signalled)
-						signalled = _transactionWaitHandle.WaitOne(DefaultTimeout);
+					lock (_transactionLock)
+					{
+						while (_transactions.Count < count && signalled)
+							signalled = _transactionWaitHandle.WaitOne(DefaultTimeout);
+					}
 				}
 				else
 				{
 					var stopWatch = Stopwatch.StartNew();
-					while (_transactions.Count < count && signalled)
+					lock (_transactionLock)
```

Comment:
If I understand this correctly, the lock around waiting for `x` count of transactions can't succeed because adding transactions is also within a lock, meaning no transactions can be added whilst waiting for the count to be reached.

I _think_ the locks around `WaitForX()` count can be removed, and that locking is needed only on adding items or copying to an immutable collection.

Reply:
Oh, there is an `AutoResetEvent.WaitOne()` within the lock 🤦 . I really wonder why no test failed, everything was green both locally and in CI. Yeah, I also think this should fail if first thread calls `WaitForX()` and then another one tries to add to the queue.

I think locking on the reading of `_transaction.Count` and `spans.Count` is needed - otherwise the number which the thread calling `WaitForX()` sees can be incorrect if another thread adds to the queue, but the lock should be limited to the call to `.Count`, and that finishes immediately.

I updated the PR accordingly.